src/preact/use-is-intersecting.js

```diff
@@ -0,0 +1,47 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../log';
+import {useEffect, useState} from './index';
+
+/** @type {null|IntersectionObserver} */ let sharedObserver = null;
+
+/**
+ * @param {{current: HTMLElement}} ref
+ * @return {boolean}
+ */
+export function useIsIntersecting(ref) {
+  const {0: isIntersecting, 1: setIsIntersecting} = useState(null);
+  if (sharedObserver === null) {
+    sharedObserver = new IntersectionObserver(entries => {
```

Comment:
We can't do this in the body of the function. Has to be inside the effect itself.

Reply:
Why does this have to be inside the effect? I recall `setIsIntersecting` is guaranteed to be stable, so I am not sure what dependencies require this to be in an effect.

However I did move this to global scope recently since [moving the setters to a map](https://github.com/ampproject/amphtml/pull/26507#discussion_r382409562), would appreciate your thoughts here.

Reply:
Specifically `new IntersectionObserver()` must be fully done inside the effect. This is a non-rendering instruction and it's not always available in different environments.